awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -440,53 +459,113 @@ def delete_if_needed(self, existing_item, handle_response=True, on_delete=None):
-                elif 'json' in response:
-                    # This is from a project delete (if there is an active job against it)
-                    if 'error' in response['json']:
-                        self.fail_json(msg="Unable to delete {0} {1}: {2}".format(item_type, item_name, response['json']['error']))
+                self.fail_json(msg="Unable to delete {0} {1}: {2}".format(item_type, item_name, self.extract_errors_from_response(response)), **{'response': response})
+        else:
+            self.exit_json(**self.json_output)
+
+    def modify_associations(self, association_endpoint, new_association_list):
+        # if we got None instead of [] we are not modifying the association_list
+        if new_association_list == None:
+            return
+
+        # First get the existing associations
+        response = self.get_all_endpoint(association_endpoint)
+        existing_associated_ids = [association['id'] for association in response['json']['results']]
+
+        # Disassociate anything that is in existing_associated_ids but not in new_association_list
+        ids_to_remove = list(set(existing_associated_ids) - set(new_association_list))
+        for an_id in ids_to_remove:
+            response = self.post_endpoint(association_endpoint, **{'data': {'id': int(an_id), 'disassociate': True}})
+            if response['status_code'] == 204:
+                self.json_output['removed_an_association'] = True
+                self.json_output['changed'] = True
+            else:
+                self.fail_json(msg="Failed to disassociate item {0}".format(self.extract_errors_from_response(response)))
+
+        # Associate anything that is in new_association_list but not in `association`
+        for an_id in list(set(new_association_list) - set(existing_associated_ids)):
+            response = self.post_endpoint(association_endpoint, **{'data': {'id': int(an_id)}})
+            if response['status_code'] == 204:
+                self.json_output['added_an_association'] = True
+                self.json_output['changed'] = True
+            else:
+                self.fail_json(msg="Failed to associate item {0}".format(self.extract_errors_from_response(response)))
+
+    def process_additional_posts(self, item, item_type, additional_posts):
+        # Process any additional posts with this item
+        for post in additional_posts:
+            endpoint = item.get('related', {}).get(post['endpoint_reference'], None)
+            if endpoint:
+                # First get the existing data from the endpoint
+                existing_data = self.get_endpoint(endpoint)
+
+                # If the endpont data is identical, just continue on to the next item
+                if existing_data['json'] == post['data']:
+                    continue
+
+                with open('/tmp/john', 'w') as f:
+                    import json
+                    f.write(json.dumps(post['data']))
```

Comment:
Not going to lie, when I'm in these situations I `raise Exception(data)` and run the test

Reply:
Yea, I should probably do something better but exceptions in Ansible can be a tad ugly to go through. Especially just when I want to try and dump a response data out. Removed the debug.